### DISCUSSION TITLE
Authelia OpenID Connect - "Unable to login via OAuth" after consentment

### DISCUSSION THREAD

**Discussion by sebastien4444** [authelia/authelia#7053] [Q&A]
Hi !

I configured Authelia (https://authelia.mydomain.com) to active openID for my portainer (https://admdocker.mydomain.com).

Here is my Authelia configuration.yaml

`
identity_providers:
  oidc:
    hmac_secret: MYHMACSECRET'
    jwks:
    -
      key: |
        -----BEGIN PRIVATE KEY-----
         MYPRIVATEKEY
        -----END PRIVATE KEY-----

    enable_client_debug_messages: false
    minimum_parameter_entropy: 8
    enforce_pkce: 'public_clients_only'
    enable_jwt_access_token_stateless_introspection: false
    discovery_signed_response_alg: 'none'
    discovery_signed_response_key_id: ''

    cors:
      endpoints:
        - 'authorization'
        - 'pushed-authorization-request'
        - 'token'
        - 'revocation'
        - 'introspection'
        - 'userinfo'

    clients:
      - 
        client_id: 'admdocker'
        client_name: 'Admdocker'
        client_secret: 'MYCLIENTSECRET'  # The digest of 'insecure_secret'.
        public: false
        authorization_policy: 'one_factor'
        redirect_uris:
          - 'https://admdocker.mydomain.com'
        scopes:
          - 'openid'
          - 'profile'
          - 'groups'
          - 'email'
        userinfo_signed_response_alg: 'none'

    lifespans:
      access_token: '1 hour'
      authorize_code: '1 minute'
      id_token: '1 hour'
      refresh_token: '90 minutes'
`

With that, I can use "Login with OAuth" button in portainer, then I authenticate through Authelia portal, I get the Authelia consentment page and click Accept but I get a "Failure Unauthorized" message from portainer and get back to the login page with "Unable to login via OAuth" message.

In my Authelia log, I don't understand what is going wrong :

`
time="2024-03-31T16:04:50Z" level=debug msg="Check authorization of subject username= groups= ip=192.168.1.254 and object https://admdocker.mydomain.com/ (method GET)."
time="2024-03-31T16:05:15Z" level=debug msg="Authorization Request with id '798ab50c-86bb-48ec-84f3-aeb6d12f8713' on client with id 'admdocker' is being processed" method=GET path=/api/oidc/authorization remote_ip=192.168.1.254
time="2024-03-31T16:05:50Z" level=debug msg="Mark 1FA authentication attempt made by user 'admin'" method=POST path=/api/firstfactor remote_ip=192.168.1.254
time="2024-03-31T16:05:50Z" level=debug msg="Successful 1FA authentication attempt made by user 'admin'" method=POST path=/api/firstfactor remote_ip=192.168.1.254
time="2024-03-31T16:05:50Z" level=debug msg="Authorization Request with id 'bb6171cb-fc8a-4ef9-9f66-0dfc7a6416c1' on client with id 'admdocker' is being processed" method=GET path=/api/oidc/authorization remote_ip=192.168.1.254
time="2024-03-31T16:05:50Z" level=debug msg="Authorization Request with id 'bb6171cb-fc8a-4ef9-9f66-0dfc7a6416c1' on client with id 'admdocker' using consent mode 'explicit' proceeding to generate a new consent session" method=GET path=/api/oidc/authorization remote_ip=192.168.1.254
time="2024-03-31T16:05:50Z" level=debug msg="Authorization Request with id 'bb6171cb-fc8a-4ef9-9f66-0dfc7a6416c1' on client with id 'admdocker' using consent mode 'explicit' authentication level 'one_factor' is sufficient for client level 'one_factor'" method=GET path=/api/oidc/authorization remote_ip=192.168.1.254
time="2024-03-31T16:05:50Z" level=debug msg="Authorization Request with id 'bb6171cb-fc8a-4ef9-9f66-0dfc7a6416c1' on client with id 'admdocker' using consent mode 'explicit' is being redirected to 'https://authelia.mydomain.com/consent?id=79d3ea72-e9c3-45e6-8525-76f014254e8a'" method=GET path=/api/oidc/authorization remote_ip=192.168.1.254
time="2024-03-31T16:07:00Z" level=debug msg="Authorization Request with id 'f082d504-40ad-442f-89e7-c28881b446b9' on client with id 'admdocker' is being processed" method=GET path=/api/oidc/authorization remote_ip=192.168.1.254
time="2024-03-31T16:07:00Z" level=debug msg="Authorization Request with id 'f082d504-40ad-442f-89e7-c28881b446b9' on client with id 'admdocker' was successfully processed, proceeding to build Authorization Response" method=GET path=/api/oidc/authorization remote_ip=192.168.1.254
time="2024-03-31T16:07:00Z" level=debug msg="Check authorization of subject username=admin groups=admins,dev ip=192.168.1.254 and object https://admdocker.mydomain.com/?code=authelia_ac_XYf9NYYdUYgUFzdsdwdSOUsp3-fUO8aCpcAO4Z0jqTM.NS0WsbOx3nhiigmbQe2X1yhNaPWJu_QSjf_YrnR1QJA&iss=https%3A%2F%2Fauthelia.mydomain.com&scope=openid+profile+groups+email&state=de4b073b-617d-487d-a1b6-6472443fbfc5 (method GET)."
time="2024-03-31T16:07:01Z" level=debug msg="Access Request with id 'f082d504-40ad-442f-89e7-c28881b446b9' on client with id 'admdocker' is being processed" method=POST path=/api/oidc/token remote_ip=192.168.1.254
time="2024-03-31T16:07:01Z" level=debug msg="Access Request with id 'f082d504-40ad-442f-89e7-c28881b446b9' on client with id 'admdocker' has successfully been processed" method=POST path=/api/oidc/token remote_ip=192.168.1.254
time="2024-03-31T16:07:01Z" level=debug msg="UserInfo Request with id 'bb2270f5-53fc-489d-8e39-9e6002c18dba' is being processed" method=GET path=/api/oidc/userinfo remote_ip=192.168.1.254
time="2024-03-31T16:07:01Z" level=debug msg="UserInfo Request with id 'bb2270f5-53fc-489d-8e39-9e6002c18dba' on client with id 'admdocker' is being returned unsigned as per the registered client configuration" method=GET path=/api/oidc/userinfo remote_ip=192.168.1.254
time="2024-03-31T16:07:01Z" level=debug msg="UserInfo Request with id 'bb2270f5-53fc-489d-8e39-9e6002c18dba' on client with id 'admdocker' was successfully processed" method=GET path=/api/oidc/userinfo remote_ip=192.168.1.254

`

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

Based on the information provided and the logs from Authelia, it seems that the authentication process through Authelia is successful, including the authorization request, consent, and user info retrieval. The issue appears to be on the Portainer side, specifically after the OAuth flow is completed and control is handed back to Portainer with the authorization code.

Given the successful OAuth flow indicated by the Authelia logs, a few potential issues on the Portainer side could lead to the "Failure Unauthorized" message:

1. **User Mapping**: Ensure that Portainer is correctly mapping the user information received from Authelia. This includes verifying that the claims (such as username or email) provided by Authelia are correctly used by Portainer to authenticate the user. The user identifier needs to be `preferred_username` or one of the other available claims as per the integration guide. This is a common issue as seen in the discussion on GitHub regarding issues with Portainer and OAuth authentication ([source](https://github.com/authelia/authelia/discussions/4298)).

2. **Redirect URI Mismatch**: The error message "The 'redirect_uri' parameter does not match any of the OAuth 2.0 Client's pre-registered redirect urls" from another discussion ([source](https://github.com/authelia/authelia/discussions/5531)) suggests that a common issue is a mismatch between the `redirect_uri` used by Portainer and the one registered in Authelia's configuration. Double-check that the `redirect_uris` in Authelia's configuration exactly matches the one used by Portainer, including the protocol (https), domain, and any path.

3. **Client Secret**: Ensure that the client secret used by Portainer matches the one configured in Authelia. Any discrepancy here could cause authentication failures.

4. **Portainer Configuration**: Review the Portainer configuration for any misconfigurations, especially around the OAuth/OpenID Connect settings. Ensure that the endpoints for authorization, token, and userinfo in Portainer match those provided by Authelia.

5. **Logs and Debugging**: Check Portainer logs for any specific error messages or hints that could point to the cause of the issue. The logs might provide more context on why the authentication is failing after the OAuth flow.

If after checking these potential issues the problem persists, consider enabling more verbose logging on both Authelia and Portainer (if not already done) to get more detailed information about the failure. Additionally, reviewing the specific OAuth/OpenID Connect settings in Portainer against Authelia's documentation and ensuring compatibility could help identify configuration mismatches.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Comment by james-d-elliott** [comment]
The authelia logs indicate it was a 100% success. Do the portainer logs show additional info?

**Comment by sebastien4444** [comment]
My mistake, I realized I have replaced "preferred_user" by my own username "admin" in Portainer Authentication config.
